### PR TITLE
PLAT-126934: Support scroll by hover

### PR DESCRIPTION
**Pull request by 0x64**
Enact-DCO-1.0-Signed-off-by: Seungcheon Baek (sc.baek@lge.com)

### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [x] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
New UX requirement to scroll a list by hover.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
Implemented a new feature.

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

### Links
[//]: # (Related issues, references)
PLAT-126934
enactjs/enact/pull/2925

### CODE REVIEW

File: Scroller/Scroller.js [hunk 1]
```diff
@@ -217,6 +219,15 @@ Scroller.propTypes = /** @lends sandstone/Scroller.Scroller.prototype */ {
 	 */
 	horizontalScrollThumbAriaLabel: PropTypes.string,
 
+	/**
+	 * Enables scroll by hover on edges
+	 *
+	 * @type {Boolean}
+	 * @default false
```
Comment (MikyungKim) [inline Scroller/Scroller.js:226]: Let's remove it.

File: Scroller/Scroller.js [hunk 2]
```diff
@@ -402,6 +413,7 @@ Scroller.defaultProps = {
 	fadeOut: false,
 	focusableScrollbar: false,
 	horizontalScrollbar: 'auto',
+	hoverToScroll: false,
```
Comment (MikyungKim) [inline Scroller/Scroller.js:416]: Let's remove this too.

File: VirtualList/VirtualList.js [hunk 3]
```diff
@@ -188,6 +190,15 @@ VirtualList.propTypes = /** @lends sandstone/VirtualList.VirtualList.prototype *
 	 */
 	horizontalScrollThumbAriaLabel: PropTypes.string,
 
+	/**
+	 * Enables scroll by hover on edges
+	 *
+	 * @type {Boolean}
+	 * @default false
```
Comment (MikyungKim) [inline VirtualList/VirtualList.js:197]: ditto.

File: useScroll/useScroll.js [hunk 4]
```diff
@@ -281,6 +286,22 @@ const useThemeScroll = (props, instances) => {
 	};
 };
 
+const HoverArea = ({direction, position, hoverToScrollStart, hoverToScrollEnd}) => (
```
Comment (MikyungKim) [inline useScroll/useScroll.js:289]: Let's make `useHoverToScroll.js` for related logic individually.

**Review by MikyungKim — CHANGES_REQUESTED** [review]
I found that focus is disappeared when pressing 5way up or down to the scrolling direction by hover. If you can fix this in this PR, please do. If not, let's create another ticket.

File: useScroll/HoverToScroll.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+ * A hover area to scroll for a single direction
+ *
+ * @class HoverToScrollBase
+ * @memberof sandstone/useScroll/HoverToScroll
+ * @ui
+ * @private
+ */
+const HoverToScrollBase = (props) => {
+	const {direction, scrollContainerHandle} = props;
+	const bounds = scrollContainerHandle.current.getScrollBounds();
```
[the remark this answers is not in the document]
Comment (0x64) [reply useScroll/HoverToScroll.js:36]: Guarded in `HoverToScroll` before using this.

File: useScroll/HoverToScroll.module.less [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+.hoverToScroll {
+	position: absolute;
+	z-index: 10;
+	pointer-events: none;
+	:global(.spotlight-input-mouse) & {
+		pointer-events: auto;
+	}
+	&.vertical {
+		width: 100%;
+		height: @sand-scroll-hover-scroll-width;
```
[the remark this answers is not in the document]
Comment (0x64) [reply useScroll/HoverToScroll.module.less:15]: 👍

**Comment by 0x64** [comment]
> I found two issues.
> 
> 1. focus is disappeared when pressing 5way up or down to the scrolling direction by hover.
> 2. The reproducibility is very rare but the focus doesn't blur when scrolling by hover.
>    If you can fix this in this PR, please do. If not, let's create another ticket.

The first issue looks like an issue of asynchronous scroll position between threads inside renderer as it is not reproducible in translate mode and the symptom is a kind of delayed stopping.

The second issue is rarely reproduced in a PC environment, and it looks not possible to happen in a real device as the reproducing steps require a very rapid input type changing like pointing-key-pointing which is not possible in a target device.

File: useScroll/HoverToScroll.js [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+									)
+								},
+								animate: false
+							});
+							startRaf(scrollByHover);
+						} else {
+							stopRaf(); // for other type input during hovering
+						}
+					};
+					startRaf(scrollByHover);
```
Comment (jeonghee27) [inline useScroll/HoverToScroll.js:124]: I didn't find any bugs about the behavior, but is it intended that startRaf() is called twice?

File: useScroll/HoverToScroll.js [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+		if (typeof window === 'object' && mutableRef.current.hoverToScrollRafId !== null) {
+			window.cancelAnimationFrame(mutableRef.current.hoverToScrollRafId);
+			mutableRef.current.hoverToScrollRafId = null;
+			if (typeof document === 'object') {
+				document.removeEventListener('keydown', handleGlobalKeyDown, {capture: true});
+			}
+		}
+	}, [handleGlobalKeyDown]);
+
+	const scrollJob = useCallback((position) => {
```
Comment (jeonghee27) [inline useScroll/HoverToScroll.js:96]: How about `onPointerEnter` instead of scrollJob

**Review by jeonghee27 — APPROVED** [review]
LGTM

File: Scroller/Scroller.js [hunk 9]
```diff
@@ -217,6 +220,14 @@ Scroller.propTypes = /** @lends sandstone/Scroller.Scroller.prototype */ {
 	 */
 	horizontalScrollThumbAriaLabel: PropTypes.string,
 
+	/**
+	 * Enables scroll by hover on edges
```
Comment (MikyungKim) [inline Scroller/Scroller.js:224]: How about
```suggestion
	 * Enables scroll by hover on edges in scroll direction.
```

File: useScroll/HoverToScroll.js [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+		before: 'down',
+		after: 'up'
+	}
+};
+
+/**
+ * A hover area to scroll for a single direction
+ *
+ * @class HoverToScrollBase
+ * @memberof sandstone/useScroll/HoverToScroll
```
Comment (MikyungKim) [inline useScroll/HoverToScroll.js:43]: Did you mean `sandstone/useScroll.HoverToScroll`?

File: useScroll/HoverToScroll.js [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+		after: 'left'
+	},
+	vertical: {
+		before: 'down',
+		after: 'up'
+	}
+};
+
+/**
+ * A hover area to scroll for a single direction
```
Comment (MikyungKim) [inline useScroll/HoverToScroll.js:40]:
```suggestion
 * A hover area to scroll for a single direction.
```

File: useScroll/HoverToScroll.js (hunk 10 above)
Comment (0x64) [reply useScroll/HoverToScroll.js:43]: Right. I will fix this.

File: useScroll/HoverToScroll.js (hunk 11 above)
Comment (0x64) [reply useScroll/HoverToScroll.js:40]: I missed two dots. I will fix them.

File: useScroll/HoverToScroll.js [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+	const mutableRef = useRef({
+		hoverToScrollRafId: null,
+		hoveredPosition: null
+	});
+
+	// Functions
+
+	const handleGlobalKeyDown = useCallback((ev) => {
+		const keyCode = ev.keyCode;
```
Comment (MikyungKim) [inline useScroll/HoverToScroll.js:61]: Either 
```suggestion
		const {keyCode} = ev;
```
or
({keyCode}) from line 60.

File: useScroll/HoverToScroll.js [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+
+import {getLastInputType} from '../ThemeDecorator';
+
+import css from './HoverToScroll.module.less';
+
+const nop = () => {};
+const propertyNames = (direction) => {
+	return direction === 'vertical' ? {
+		axis: 'y',
+		client: 'clientHeight',
```
Comment (MikyungKim) [inline useScroll/HoverToScroll.js:17]: how about
```suggestion
		clientSize: 'clientHeight',
```

**Review by MikyungKim — COMMENTED** [review]
I think we have ui test failure from this branch. Please take a look.
```
 1) with spottable should set correct focus after Page Up [ENYO-6281]
[chrome 79.0.3945.88 Linux #0-25] button 1 has focus again: expected false to be true
[chrome 79.0.3945.88 Linux #0-25] AssertionError: button 1 has focus again: expected false to be true
[chrome 79.0.3945.88 Linux #0-25]     at Context.<anonymous> (/home/jenkins/workspace/enact-ui-tests/sandstone/tests/ui/specs/Scroller/WithSpottable/ScrollerWithSpottable-specs.js:27:83)
```

**Comment by 0x64** [comment]
> I think we have ui test failure from this branch. Please take a look.
> 
> ```
>  1) with spottable should set correct focus after Page Up [ENYO-6281]
> [chrome 79.0.3945.88 Linux #0-25] button 1 has focus again: expected false to be true
> [chrome 79.0.3945.88 Linux #0-25] AssertionError: button 1 has focus again: expected false to be true
> [chrome 79.0.3945.88 Linux #0-25]     at Context.<anonymous> (/home/jenkins/workspace/enact-ui-tests/sandstone/tests/ui/specs/Scroller/WithSpottable/ScrollerWithSpottable-specs.js:27:83)
> ```

Good catch! Thank you. It was an issue of UI test app which uses two classes having a conflict CSS attribute value at the same time. In this case, the definition order of classes can affect the result.

**Review by MikyungKim — APPROVED** [review]
LGTM

**Pull request merged**